src/Microsoft.Net.Http.Headers/HttpHeaderParser.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Globalization;
+
+namespace Microsoft.Net.Http.Headers
+{
+    internal abstract class HttpHeaderParser<T>
+    {
+        private bool _supportsMultipleValues;
+
+        protected HttpHeaderParser(bool supportsMultipleValues)
+        {
+            _supportsMultipleValues = supportsMultipleValues;
+        }
+
+        public bool SupportsMultipleValues
+        {
+            get { return _supportsMultipleValues; }
+        }
+
+        // If a parser supports multiple values, a call to ParseValue/TryParseValue should return a value for 'index'
+        // pointing to the next non-whitespace character after a delimiter. E.g. if called with a start index of 0
+        // for string "value , second_value", then after the call completes, 'index' must point to 's', i.e. the first
+        // non-whitespace after the separator ','.
+        public abstract bool TryParseValue(string value, ref int index, out T parsedValue);
+
+        public T ParseValue(string value, ref int index)
+        {
+            // Index may be value.Length (e.g. both 0). This may be allowed for some headers (e.g. Accept but not
+            // allowed by others (e.g. Content-Length). The parser has to decide if this is valid or not.
+            Contract.Requires((value == null) || ((index >= 0) && (index <= value.Length)));
+
+            // If a parser returns 'null', it means there was no value, but that's valid (e.g. "Accept: "). The caller
+            // can ignore the value.
+            T result = default(T);
+            if (!TryParseValue(value, ref index, out result))
+            {
+                throw new FormatException(string.Format(CultureInfo.InvariantCulture, "Invalid value '{0}'.",
+                    value == null ? "<null>" : value.Substring(index)));
+            }
+            return result;
+        }
+
+        public virtual bool TryParseValues(IList<string> values, out IList<T> parsedValues)
```

Comment:
This doesn't require an out parameter since you're returning a reference type (IList) instead of a value type.  You could just return the parsed values directly.